### PR TITLE
Panda, tiger, eagle!

### PR DESCRIPTION
**Pull request by noahpatterson**
A few notes/questions:
- need help simplifying the specs with class variables
- noticed another student used self in the initialize, that would have simplified things
- added functionality to also update the class variable on each update
- added functionality to allow filter by any combination of attributes

### CODE REVIEW

File: models/motorcycle.rb [hunk 1]
```diff
@@ -0,0 +1,7 @@
+require_relative 'vehicle'
+
+class Motorcycle < Vehicle
+  def self.num_wheels
+    num_wheels = 2
```
Comment (jwo) [inline models/motorcycle.rb:5]: this is the same as

``` ruby
def self.num_wheels
  2
end
```

File: models/vehicle.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
+class Vehicle
+
+  attr_reader :color, :make, :model, :year, :name
+  
+  @@all_vehicles = []
+
+  def initialize(args={})
+    @color = args.fetch(:color) if args[:color]
```
Comment (jwo) [inline models/vehicle.rb:8]: if you're going to use `if args[:color]`, that means it's ok if `@color` is nil.

If that's the case, this syntax is preferred:

``` ruby
@color = args[:color]
```

If you want a default value, you can do this:

``` ruby
@color = args.fetch(:color, "red")
```

File: models/vehicle.rb [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  def self.num_wheels
+    num_wheels = 4
+  end
+
+  def self.filter(args ={})
+    search_str =  args.to_a.map { |group| "vehicle[:#{group[0]}] == '#{group[1]}' and "}.reduce(:+)[0..-5]
+    @@all_vehicles.find_all { |vehicle| eval(search_str) }
+  end
+
+  def update_all_vehicles(name)
```
Comment (jwo) [inline models/vehicle.rb:30]: Something that can update all instances of a class should be on the class itself. That is, the method should be "scoped" to the object -- it should have the same powers as the object itself.

tl;dr

If it can update more than 1 thing:

``` ruby
def self.update_all_vehicles(name)
```

File: models/vehicle.rb [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  def all_vehicles
+    @@all_vehicles
+  end
+
+  def update(name, args={})
+    @color = args.fetch(:color) if args[:color]
+    @make = args.fetch(:make) if args[:make]
+    @model = args.fetch(:model) if args[:model]
+    @year = args.fetch(:year).to_s if args[:year]
+    update_all_vehicles(name)
```
Comment (jwo) [inline models/vehicle.rb:44]: This went a bit beyond the spec --- it makes sense to do this, but it wasn't required of you. 

Mostly, if I have a vehicle, I want to be able to update it.

``` ruby
class Vehicle
  def update(args)
    @color = args.fetch(:color) if args[:color]
    @make = args.fetch(:make) if args[:make]
    @model = args.fetch(:model) if args[:model]
    @year = args.fetch(:year).to_s if args[:year]
    self
  end
end
```

File: models/vehicle.rb [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    num_wheels = 4
+  end
+
+  def self.filter(args ={})
+    search_str =  args.to_a.map { |group| "vehicle[:#{group[0]}] == '#{group[1]}' and "}.reduce(:+)[0..-5]
+    @@all_vehicles.find_all { |vehicle| eval(search_str) }
+  end
+
+  def update_all_vehicles(name)
+    index = @@all_vehicles.index {|vehicle| vehicle[:name] == name}
```
Comment (jwo) [inline models/vehicle.rb:31]: so, instead of using "index", in ruby you have the "find" method, and then update it in place.

``` ruby
  Vehicle.all_vehicles.find{|v| v.name == name}.update(args)
```

Example usage: http://rubyfiddle.com/riddles/37f83

**Comment by noahpatterson** [comment]
Thanks for the feedback. I might try to work on a v2 of this at a later date to see I can grasp how to save the objects to @@all_vehicles.
